clang_build/project.py

```diff
@@ -0,0 +1,300 @@
+'''
+Target describes a single build or dependency target with all needed paths and
+a list of buildables that comprise it's compile and link steps.
+'''
+
+import os as _os
+import textwrap as _textwrap
+import sys
+from pathlib import Path as _Path
+import subprocess as _subprocess
+from multiprocessing import freeze_support as _freeze_support
+import logging as _logging
+
+import toml
+from .dialect_check import get_max_supported_compiler_dialect as _get_max_supported_compiler_dialect
+from .build_type import BuildType as _BuildType
+from .target import Executable as _Executable,\
+                    SharedLibrary as _SharedLibrary,\
+                    StaticLibrary as _StaticLibrary,\
+                    HeaderOnly as _HeaderOnly
+from .dependency_tools import find_circular_dependencies as _find_circular_dependencies,\
+                              find_non_existent_dependencies as _find_non_existent_dependencies,\
+                              get_dependency_walk as _get_dependency_walk
+from .io_tools import get_sources_and_headers as _get_sources_and_headers
+from .progress_bar import CategoryProgress as _CategoryProgress,\
+                          IteratorProgress as _IteratorProgress
+from .logging_stream_handler import TqdmHandler as _TqdmHandler
+
+_LOGGER = _logging.getLogger('clang_build.clang_build')
+
+
+
+class Project:
+    def __init__(self, config, environment, multiple_projects):
+
+        self.name = config.get("name", "")
+
+        # print(f"Project {self.name}")
+        self.workingdir = environment.workingdir
+
+        # Project build directory
+        self.build_directory = environment.build_directory
+        if multiple_projects:
+            self.build_directory = self.build_directory.joinpath(self.name)
+
+        ### TODO: external sources should be fetched before any sources are read in, i.e. even before projects are created
```

Comment:
Is this TODO still open? Is it important for sub-projects?

Reply:
For projects it's fine as is, I removed the comment with 01e51cab00b7badd77fbb00b776fb2b0cb3371e5.